fuel-tx/src/transaction/types/upgrade.rs

```diff
@@ -0,0 +1,315 @@
+                // Nothing metadata for state transition upgrades.
+                Ok(Self::StateTransition)
+            }
+        }
+    }
+}
+
+/// The types describe the purpose of the upgrade performed by the [`Upgrade`]
+/// transaction.
+#[derive(Copy, Clone, Derivative, strum_macros::EnumCount)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(fuel_types::canonical::Deserialize, fuel_types::canonical::Serialize)]
+#[derivative(Eq, PartialEq, Hash, Debug)]
+pub enum UpgradePurpose {
+    /// The upgrade is performed to change the consensus parameters.
+    ConsensusParameters {
+        /// The index of the witness in the [`Witnesses`] field that contains
+        /// the serialized consensus parameters.
+        witness_index: u16,
+        /// The hash of the serialized consensus parameters.
+        /// Since the serialized consensus parameters live inside witnesses(malleable
+        /// data), any party can override them. The `checksum` is used to verify that the
+        /// data was not modified.
+        checksum: Bytes32,
+    },
+    /// The upgrade is performed to change the state transition function.
+    StateTransition {
+        /// The hash of the new bytecode of the state transition function.
+        /// The bytecode must be present on the blockchain(should be known by the
+        /// network) at the moment of inclusion of this transaction.
+        bytecode_hash: Bytes32,
+    },
+}
+
+/// The body of the [`Upgrade`] transaction.
+#[derive(Clone, Derivative)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[derive(fuel_types::canonical::Deserialize, fuel_types::canonical::Serialize)]
+#[canonical(prefix = TransactionRepr::Script)]
+#[derivative(Eq, PartialEq, Hash, Debug)]
+pub struct UpgradeBody {
+    /// The purpose of the upgrade.
+    pub(crate) purpose: UpgradePurpose,
+}
+
+impl Default for UpgradeBody {
+    fn default() -> Self {
+        Self {
+            purpose: UpgradePurpose::StateTransition {
+                bytecode_hash: Default::default(),
+            },
+        }
+    }
+}
+
+impl PrepareSign for UpgradeBody {
+    fn prepare_sign(&mut self) {}
+}
+
+impl Chargeable for Upgrade {
+    #[inline(always)]
+    fn metered_bytes_size(&self) -> usize {
+        Serialize::size(self)
+    }
+
+    #[inline(always)]
+    fn gas_used_by_metadata(&self, gas_cost: &GasCosts) -> Word {
+        let bytes = Serialize::size(self);
+        // Gas required to calculate the `tx_id`.
+        let tx_id_gas = gas_cost.s256().resolve(bytes as u64);
+
+        let purpose_gas = match &self.body.purpose {
+            UpgradePurpose::ConsensusParameters { witness_index, .. } => {
+                let len = self
+                    .witnesses
+                    .get(*witness_index as usize)
+                    .map_or(0, |w| w.as_vec().len());
+                gas_cost.s256().resolve(len as u64)
+            }
```

Comment:
I'm not sure why we charge for this. Having a spec update PR would help.

Reply:
We charge for checksum calculation. I will update the specification with the `Upgrade` transaction soon

Reply:
https://github.com/FuelLabs/fuel-specs/pull/567